RecoMuon/SimMuonAlgos/plugins/MuonSimClassifier.cc

```diff
@@ -0,0 +1,574 @@
+
+//
+// class decleration
+class MuonSimClassifier : public edm::EDProducer {
+    public:
+        explicit MuonSimClassifier(const edm::ParameterSet&);
+        ~MuonSimClassifier() override;
+
+    private:
+        void produce(edm::Event&, const edm::EventSetup&) override;
+        /// The RECO objects
+        edm::EDGetTokenT<edm::View<reco::Muon> > muonsToken_;
+
+        /// Track to use
+        reco::MuonTrackType trackType_;
+
+        /// The TrackingParticle objects
+        edm::EDGetTokenT<TrackingParticleCollection> trackingParticlesToken_;
+
+        /// The Associations
+        edm::InputTag associatorLabel_;
+        edm::EDGetTokenT<reco::MuonToTrackingParticleAssociator> muAssocToken_;
+
+        /// Cylinder to use to decide if a decay is early or late
+        double decayRho_, decayAbsZ_;
+
+        /// Create a link to the generator level particles
+        bool linkToGenParticles_;
+        edm::InputTag genParticles_;
+        edm::EDGetTokenT<reco::GenParticleCollection> genParticlesToken_;
+
+        /// Returns the flavour given a pdg id code
+        int flavour(int pdgId) const ;
+
+        /// Write a ValueMap<int> in the event
+        template<typename T>
+        void writeValueMap(edm::Event &iEvent,
+                const edm::Handle<edm::View<reco::Muon> > & handle,
+                const std::vector<T> & values,
+                const std::string    & label) const ;
+
+        TrackingParticleRef getTpMother(TrackingParticleRef tp) {
+            if (tp.isNonnull() && tp->parentVertex().isNonnull() && !tp->parentVertex()->sourceTracks().empty()) {
+                return tp->parentVertex()->sourceTracks()[0];
+            } else {
+                return TrackingParticleRef();
+            }
+        }
+
+        /// Convert TrackingParticle into GenParticle, save into output collection,
+        /// if mother is primary set reference to it,
+        /// return index in output collection
+        int convertAndPush(const TrackingParticle &tp,
+                           reco::GenParticleCollection &out,
+                           const TrackingParticleRef &momRef,
+                           const edm::Handle<reco::GenParticleCollection> & genParticles) const ;
+};
+
+
+MuonSimClassifier::MuonSimClassifier(const edm::ParameterSet &iConfig) :
+    muonsToken_(consumes<edm::View<reco::Muon> >(iConfig.getParameter<edm::InputTag>("muons"))),
+    trackingParticlesToken_(consumes<TrackingParticleCollection>(iConfig.getParameter<edm::InputTag>("trackingParticles"))),
+    muAssocToken_(consumes<reco::MuonToTrackingParticleAssociator>(iConfig.getParameter<edm::InputTag>("associatorLabel"))),
+    decayRho_(iConfig.getParameter<double>("decayRho")),
+    decayAbsZ_(iConfig.getParameter<double>("decayAbsZ")),
+    linkToGenParticles_(iConfig.getParameter<bool>("linkToGenParticles")),
+    genParticles_(linkToGenParticles_ ? iConfig.getParameter<edm::InputTag>("genParticles") : edm::InputTag("NONE"))
```

Comment:
an empty input tag can be used instead of "NONE"

Reply:
fixed